macros/core/PGauxiliaryFunctions.pl

```diff
@@ -532,5 +539,45 @@ sub fact {
 	P($_[0], $_[0]);
 }
 
+=head3 random_subset function
+
+=pod
+
+	Usage: random_subset($n, @set);
+
+This function returns a randomly ordered array of $n elements selected from the array @set
+without replacement. Accepts either an array or an array reference for @set.
+
+Example to choose 3 random elements (both do the same thing):
+
+	random_subset(3, 'first', 'second', 'third', 'fourth', 'fifth')
+	random_subset(3, ['first', 'second', 'third', 'fourth', 'fifth'])
+
+=cut
+
+sub random_subset {
+	my ($n, @set) = @_;
+	@set = @{ $set[0] } if (scalar(@set) == 1 && ref($set[0]) eq 'ARRAY');
+
+	unless ($n =~ /^\d+/) {
+		warn 'random_subset: The first input must be a non-negative integer';
+		return;
+	}
+	unless (@set) {
+		warn 'random_subset: The list of elements to choose from must contain at least one element.';
+		return;
+	}
+	if ($n > scalar(@set)) {
+		warn 'random_subset: The first input must be smaller than or equal to the number of elements.';
+		$n = scalar(@set);
+	}
+
+	my @out;
+	while (@set && @out < $n) {
+		push(@out, splice(@set, random(0, $#set, 1), 1));
+	}
+	return @out;
```

Comment:
Maybe this should be:

```
return wantarray ? @out : \@out;
```

Reply:
To me that seems like it would require another input `$wantarray` to the function, which I don't think is worth adding for that.

Reply:
Sorry, my mistake, this is a perl internal, and seems like a useful change. Thanks